Delegate Bitcoin Core's private key management to Eclair

sstone opened the pull request:
We want to use Eclair to manage private keys for Bitcoin Core to protect our onchain funds.
We would create empty Bitcoin Core wallets and then import descriptors generated by Eclair (this is similar to how hardware wallets are used with Bitcoin Core or Electrum).

To validate transactions without a trusted display and without human interaction, we define a very simple policy:
- we explicitly track all inputs and outputs that belong to our onchain wallet
- we only sign inputs that are tagged as belonging to us
- we verify that we can spend all outputs that are tagged as belonging to us

This works when onchain transactions are created by Eclair (funding, dual-funding, RBF, but also with the trusted swaps that we are currently using).

Bitcoin Core would treat Eclair’s onchain wallet as a watch-only wallet and will not be able to sign transactions: “sendtoaddress” and “walletprocesspsbt”  RPC call will fail. To send funds to a specific onchain address, users will need to call Eclair’s “sendonchain” API which requires human validation with a trusted Ledger device.

It would also be possible to initialise a Ledger device with Eclair’s onchain seed and keep it as a backup but using it would interfere with Eclair’s operations (both wallets may try and use the same UTXOs). 

## Wallet descriptors
Descriptors describe how to generate public keys. Bitcoin Core will use these descriptors to create a wallet, track public keys and utxos, fund transactions and provide information about its wallet inputs and outputs (like amounts being spent and BIP32 paths of the wallet public keys) which Eclair will use internally to validate transactions before signing them.

This is a descriptor for a BIP84 wallet:
```
wpkh([92aeacc2/84'/1'/0']tpubDCyGWF1tZPqSDq7HqKtM31ZDDFoYybL56RDrfKHFQXTmx1rf5Uh6T5aFTE7wZmxaCQNCzfnu6wMv53k8kZNEtmSpgiCAuvBWmv7zjBxZcAF/0/*)#hdngjcua
```
   
It includes an “xpub”, a derivation template (/0/*), and the type of script to generate (wpkh)

# Implementation

What we need to do in Eclair is:
- implement a new key manager to manage onchain keys (we’ll follow the BIP84 specs)
- add an API call to return wallet descriptors
- modify our Bitcoin Core client to use our keymanager to sign transactions 

## Signature Workflow

Whenever we need to fund and sign a transaction we:
- create an non-funded transaction
- ask bitcoin core to add inputs and outputs, with a specified fee rate
- mark all inputs and outputs that were not in the non-funded transaction as belonging to us
- create a PSBT from the funded transaction
- ask bitcoin core to add metadata (spent amounts, BIP derivations paths) to the PSBT
- check that the fees for this PSBT match the target feerate
- for all our inputs, check that we can recompute their pubkey script using the provided BIP32 paths and sign them
- for all our outputs,  check that we can recompute their pubkey script using the provided BIP32 paths

This workflow is safe because:
- we only sign inputs that we explicitly tag as belonging to us. Even if a hacker manager to control our Bitcoin Node and used our UTXOs to negotiate a dual-funding channel with us, we would not sign these inputs (because during the interactive tx negotiation Eclair will explicitly tag them as “theirs/remote”)
- we verify that we can actually spend our wallet outputs: Bitcoin Core cannot lie to us about change outputs 
- we are protected against overpaying fees because Bitcoin Core cannot lie about the amounts of the wallet inputs that are spend, because they are included in the hash that is signed (see https://github.com/bitcoin/bips/blob/master/bip-0143.mediawiki#specification)

We check that the full transaction that is spent is included in the PSBT’s non-witness utxo field. 

# Master Seed Management and Compatibility with other BIP32 wallets
We define  a new `eclair-signer.conf` file that includes BIP32 mnemonic words and passphrase. It should be compatible with any BIP32-compliant wallet. 
Descriptors generated by Eclair can also be used to create a watch-only wallet on another Bitcoin Node (we could also simply copy Eclair’s wallet since it does not contain private keys).

And a new “getmasterxpub” API call has been added to eclair, which will return an “xpub” that can be used to create an Electrum watch-only wallet.

remyers reviewed (changes requested):
So far I have only found minor issues and noted some questions. I'll look at the tests next. 

It will help with review to add something to the release notes and perhaps a new guide to `docs\` about how to setup and use the watch-only wallet when in `external_signer` mode.

sstone commented on eclair-core/src/main/scala/fr/acinq/eclair/Eclair.scala:
```diff
@@ -169,6 +169,12 @@ trait Eclair {
 
+  def getOnchainMasterPubKey(account: Long): String
+
+  def getOnchainMasterMasterFingerprintHex: String
```
fixed in https://github.com/ACINQ/eclair/pull/2613/commits/6f538f68778da2d7174007280473ec7035020bf2

sstone commented on eclair-core/src/main/scala/fr/acinq/eclair/channel/publish/ReplaceableTxFunder.scala:
```diff
@@ -308,31 +310,45 @@ private class ReplaceableTxFunder(nodeParams: NodeParams,
+
+    val psbt = new Psbt(locallySignedTx.txInfo.tx)
+    val updated = psbt.updateWitnessInput(locallySignedTx.txInfo.input.outPoint, locallySignedTx.txInfo.input.txOut, null, fr.acinq.bitcoin.Script.parse(locallySignedTx.txInfo.input.redeemScript), null, java.util.Map.of())
+    val finalized = EitherKt.flatMap(updated, (psbt: Psbt) => psbt.finalizeWitnessInput(0, locallySignedTx.txInfo.tx.txIn.head.witness))
```
Because it is the one input that eclair controls and is not managed by bitcoin core. We finalize (sign) this input and ask bitcoin core to sign the other inputs.

remyers replied:
Ok - got it. So that first input is the one for the channel controlled by the lightning key signer. Elsewhere we sign the funding inputs with the new on-chain key signer.

sstone commented on eclair-core/src/main/scala/fr/acinq/eclair/crypto/keymanager/LocalOnchainKeyManager.scala:
```diff
@@ -0,0 +1,130 @@
+
+    val input = psbt.getInput(pos)
+
+    // For each wallet input, Bitcoin Core will provide
```
I've added a test in https://github.com/ACINQ/eclair/pull/2613/commits/6f538f68778da2d7174007280473ec7035020bf2

sstone commented on eclair-core/src/main/scala/fr/acinq/eclair/crypto/keymanager/OnchainKeyManager.scala:
```diff
@@ -0,0 +1,34 @@
+   */
+  def getOnchainMasterPubKey(account: Long): String
+
+  def getOnchainMasterMasterFingerprint: Long
```
fixed in https://github.com/ACINQ/eclair/pull/2613/commits/6f538f68778da2d7174007280473ec7035020bf2

remyers replied:
There seem to be a few other instances if you search for "OnchainMasterMaster"; eg. in `OnchainKeyManager` and `LocalOnchainKeyManager`.

sstone commented on eclair-core/src/main/scala/fr/acinq/eclair/channel/publish/ReplaceableTxFunder.scala:
```diff
@@ -367,23 +385,35 @@ private class ReplaceableTxFunder(nodeParams: NodeParams,
             val packageWeight = commitTx.weight() + dummySignedTx.tx.weight()
+
+            // above, we asked bitcoin core to use the package weight to estimate fees when it built and funded this transaction, so we
+            // use the same package weight here to compute thet actual fee rate that we get
```
fixed in https://github.com/ACINQ/eclair/pull/2613/commits/6f538f68778da2d7174007280473ec7035020bf2

sstone commented on eclair-core/src/test/scala/fr/acinq/eclair/blockchain/bitcoind/BitcoinCoreClientSpec.scala:
```diff
@@ -58,25 +61,32 @@ class BitcoinCoreClientSpec extends TestKitBaseClass with BitcoindService with A
+    assume(!useExternalSigner)
+
     val sender = TestProbe()
     val bitcoinClient = new BitcoinCoreClient(bitcoinrpcclient)
```
done in https://github.com/ACINQ/eclair/pull/2613/commits/6f538f68778da2d7174007280473ec7035020bf2

sstone commented on eclair-core/src/main/scala/fr/acinq/eclair/blockchain/bitcoind/rpc/BitcoinCoreClient.scala:
```diff
@@ -389,6 +453,41 @@ class BitcoinCoreClient(val rpcClient: BitcoinJsonRPCClient) extends OnChainWall
+   * If the current wallet uses Eclair to sign transaction, then we'll use our onchain key manager to sign the transaction,
+   * with the following assumptions:
+   * - all inputs belong to us
+   * - all outputs except for the one that sends to `pubkeyScript` belong to us`
```
fixed in https://github.com/ACINQ/eclair/pull/2613/commits/6f538f68778da2d7174007280473ec7035020bf2

sstone commented on eclair-core/src/main/scala/fr/acinq/eclair/crypto/keymanager/LocalOnchainKeyManager.scala:
```diff
@@ -0,0 +1,130 @@
+
+  import LocalOnchainKeyManager._
+
+  // master key used for HWI wallets
```
I've updated the comment to explain that we are generating a master key that will be used to create BIP84-compatible wallets, using descriptors through HWI (for Bitcoin Core)  or by importing the mnemonics or the master xpub (for BIP84-compatible wallets)

remyers replied:
It's not required to use bitcoin core's HWI interface to use it as a watch-only descriptor wallet (see [comment](https://github.com/ACINQ/eclair/pull/2613#discussion_r1147572011) above).

Bitcoin Core does not access Eclair's API, but Eclair accesses Bitcoin Core via it's RPC interface to setup watch only descriptors and construct transactions. How about instead:

```
  // master key. a Bitcoin Core watch-only wallet initialized with BIP84 descriptors derived from 
  // this key can be used to construct funded unsigned transactions that Eclair can sign.
```

remyers replied:
Fixed in [53c3c77](https://github.com/ACINQ/eclair/pull/2613/commits/53c3c7733ac8d3ff88953973f9f9ecbbfcc99e40).

sstone commented on eclair-core/src/test/scala/fr/acinq/eclair/blockchain/bitcoind/BitcoindService.scala:
```diff
@@ -66,6 +72,28 @@ trait BitcoindService extends Logging {
   var bitcoincli: ActorRef = _
+  val onchainKeyManager = new LocalOnchainKeyManager(ByteVector.fromValidHex("01" * 32), Block.RegtestGenesisBlock.hash)
+
+  def setExternalSignerScript(keyManager: OnchainKeyManager): Unit = {
```
Yes we need to instantiate different fresh HWI-based bitcoin wallets in some of the tests.

remyers replied:
After we have created the watch-only descriptor wallet in `waitForBitcoindReady` couldn't we just use the `importdescriptors` RPC call instead of defining a `signer` script in `startBitcoind`?

remyers replied:
I created a test script [start_bitcoin_descriptor.sh](https://github.com/remyers/lightning-cfg/blob/descriptor-wallet/scripts/start_bitcoin_descriptor.sh) that shows how I would expect to initialize the descriptor wallet for testing without using an external signing script.

remyers replied:
Great job, fixed it in [53c3c77](https://github.com/ACINQ/eclair/pull/2613/commits/53c3c7733ac8d3ff88953973f9f9ecbbfcc99e40). Conceptually more clear now.

sstone commented on eclair-core/src/main/scala/fr/acinq/eclair/crypto/keymanager/LocalOnchainKeyManager.scala:
```diff
@@ -0,0 +1,130 @@
+
+    // update the input which the right script for a p2wpkh input, which is a * p2pkh * script
+    // then sign and finalized the psbt input
+    val updated = psbt.updateWitnessInput(psbt.getGlobal.getTx.txIn.get(pos).outPoint, input.getWitnessUtxo, null, Script.pay2pkh(pub), SigHash.SIGHASH_ALL, input.getDerivationPaths)
```
I've added a specific test in https://github.com/ACINQ/eclair/pull/2613/commits/6f538f68778da2d7174007280473ec7035020bf2

remyers commented on eclair-core/src/main/scala/fr/acinq/eclair/blockchain/bitcoind/rpc/BitcoinCoreClient.scala:
```diff
@@ -257,41 +332,30 @@ class BitcoinCoreClient(val rpcClient: BitcoinJsonRPCClient) extends OnChainWall
 
   def signTransaction(tx: Transaction, allowIncomplete: Boolean)(implicit ec: ExecutionContext): Future[SignTransactionResponse] = signTransaction(tx, Nil, allowIncomplete)
 
   def signTransaction(tx: Transaction, previousTxs: Seq[PreviousTx], allowIncomplete: Boolean = false)(implicit ec: ExecutionContext): Future[SignTransactionResponse] = {
```
I can see you fixed this in [a733b23](https://github.com/ACINQ/eclair/pull/2613/commits/a733b230280e936ac99dfa314133d73bd674a6bc)

remyers commented on eclair-core/src/main/scala/fr/acinq/eclair/blockchain/bitcoind/rpc/BitcoinCoreClient.scala:
```diff
@@ -512,6 +611,31 @@ object BitcoinCoreClient {
+    import KotlinUtils._
+
+    // Extract a fully signed transaction from `psbt`
+    // If the transaction is just partially signed, this method wil fail and you must call extractPartiallySignedTx instead
```
Fixed in [a733b23](https://github.com/ACINQ/eclair/pull/2613/commits/a733b230280e936ac99dfa314133d73bd674a6bc)

remyers reviewed (changes requested):
Changes to remove HWI look great. Especially s/`WithExternalSigner`/`WithEclairSigner` - this eliminated a major source of cognitive dissonance for me.

While looking at this some more, I noticed it should be possible to remove params from or possibly even migrated `getDescriptors` to bitcoin-kmp.

sstone commented on eclair-core/src/main/scala/fr/acinq/eclair/Eclair.scala:
```diff
@@ -169,6 +169,12 @@ trait Eclair {
+
+  def getOnchainMasterFingerprintHex: String
+
+  def getDescriptors(fingerprint: Int, chain_opt: Option[String], account: Long): (List[String], List[String])
```
I simplified our API in https://github.com/ACINQ/eclair/pull/2613/commits/9e0746b5d288954edd57c90eeba97040a316e4ab, `account` is now optional and defaults to 0

remyers replied:
Looks good!

remyers reviewed (changes requested):
We should always check that the `fee` returned in `FundTransactionResult` matches what we requested - as you do in `makeFundingTx` and `sendToPubkeyScript`. 

Because we must call `signPsbt` before we have all of the information this check can't be done in one place. This is not ideal because if we add a new call to `fundTransaction` we are likely to forget to do the check. 

Would it make sense to ***not*** return the `fee` result in `FundTransactionResult` as a way to remove the chance it could be relied on without confirming it matches what we requested?

sstone commented on eclair-core/src/main/scala/fr/acinq/eclair/blockchain/bitcoind/rpc/BitcoinCoreClient.scala:
```diff
@@ -389,6 +443,41 @@ class BitcoinCoreClient(val rpcClient: BitcoinJsonRPCClient) extends OnChainWall
+      signedTx = signedPsbt.finalTx
+      actualFees = kmp2scala(signedPsbt.psbt.computeFees())
+      actualFeerate = FeeratePerKw((actualFees * 1000) / signedTx.weight())
+      maxFeerate = actualFeerate + actualFeerate / 2
```
Good catch, done in https://github.com/ACINQ/eclair/pull/2613/commits/b55f014c52f5e6ec8f399325ec42e7b2f211aa53

remyers reviewed (changes requested):
Just found another issue with `fundrawtransaction` that looks like we can fix by adding some requires in the bitcoind client.

sstone commented on eclair-core/src/main/scala/fr/acinq/eclair/channel/publish/ReplaceableTxFunder.scala:
```diff
@@ -365,39 +391,69 @@ private class ReplaceableTxFunder(nodeParams: NodeParams,
+    }
+
+    for {
+      fundTxResponse <- bitcoinClient.fundTransaction(txNotFunded, FundTransactionOptions(targetFeerate, inputWeights = anchorWeight))
```
done in https://github.com/ACINQ/eclair/pull/2613/commits/e2953327d3636bbc82f30e98e6f28fc225775fe1

remyers commented on eclair-core/src/main/scala/fr/acinq/eclair/blockchain/bitcoind/rpc/BitcoinCoreClient.scala:
```diff
@@ -430,34 +430,48 @@ class BitcoinCoreClient(val rpcClient: BitcoinJsonRPCClient, val onchainKeyManag
+      // check that when we manage private keys we can re-compute the public key we got from bitcoin core
+      val computed_opt = this.onchainKeyManager_opt.map(_.getPublicKey(DeterministicWallet.KeyPath(keyPath)))
+      require(computed_opt.forall(_ == extracted), "cannot recompute pubkey generated by bitcoin core")
+      extracted
```
Looks good! much cleaner how you return both the pubkey and address from the onchain key manager.

sstone commented on eclair-core/src/main/scala/fr/acinq/eclair/blockchain/bitcoind/rpc/BitcoinCoreClient.scala:
```diff
@@ -223,6 +223,13 @@ class BitcoinCoreClient(val rpcClient: BitcoinJsonRPCClient, val onchainKeyManag
+      require(addedOutputs == 0 || changePos >= 0, "change output added, but position not returned")
+      require(changePos < 0 || !tx.txOut.contains(fundedTx.txOut(changePos.intValue)), "existing output returned as change output")
+      require(options.changePosition.isEmpty || changePos == options.changePosition.get || changePos == -1, "change position added at wrong position")
+
```
We already have tests for the change output in https://github.com/ACINQ/eclair/blob/70b047588f817e95c692d8051c88552b60d9ac39/eclair-core/src/test/scala/fr/acinq/eclair/blockchain/bitcoind/BitcoinCoreClientSpec.scala#L142

remyers replied:
I was thinking tests that uses a mock `BasicBitcoinJsonRPCClient` so we can create responses from `bitcoind` that it currently does not produce; ie. adds an additional output to the transaction, or a change output at a position that doesn't match the `changePos` json field. The test in `BitcoinCoreClientSpec` does not test malicious/broken `bitcoind` responses.

remyers replied:
Fixed in [205e8a3](https://github.com/ACINQ/eclair/pull/2613/commits/205e8a3f565d5fed53e506527846281d088b8776), and I learned some clever Scala techniques. :+1:

remyers commented on eclair-core/src/test/scala/fr/acinq/eclair/blockchain/bitcoind/BitcoinCoreClientSpec.scala:
```diff
@@ -140,34 +151,82 @@ class BitcoinCoreClientSpec extends TestKitBaseClass with BitcoindService with A
+      {
+        val bitcoinClient1 = makeEvilBitcoinClient((pos: Int) => 1, (tx: Transaction) => tx.copy(txOut = tx.txOut.reverse))
+        bitcoinClient1.fundTransaction(txNotFunded, FundTransactionOptions(TestConstants.feeratePerKw, changePosition = Some(0))).pipeTo(sender.ref)
+        sender.expectMsgType[Failure]
```
In the success case you check that the tx was published (rollback = true), should you also check the failure cases to ensure that wallet inputs are unlocked ?

sstone replied:
fixed in https://github.com/ACINQ/eclair/pull/2613/commits/7bb823645f3250b6ba344e4167fc7b12e099bacf

remyers commented on eclair-core/src/main/scala/fr/acinq/eclair/blockchain/bitcoind/rpc/BitcoinCoreClient.scala:
```diff
@@ -42,13 +46,47 @@ import scala.util.{Failure, Success, Try}
+  def createDescriptorWallet()(implicit ec: ExecutionContext): Future[Boolean] = {
+    onchainKeyManager_opt match {
+      case None => Future.successful(true) // no eclair-backed wallet is configured
+      case Some(onchainKeyManager) if !rpcClient.wallet.contains(onchainKeyManager.wallet) => Future.successful(true) // configured wallet has a different name
```
Should we produce a warning if a wallet exists with a different name and the onchainKeyManager is defined? The user might expect the defined descriptor wallet is used instead of the current wallet.

remyers commented on docs/BitcoinCoreKeys.md:
```diff
@@ -0,0 +1,119 @@
+You can also use `eclair-cli getmasterxpub` to get a BIP32 extended public key that you can import into any compatible Bitcoin wallet
+to create a watch-only wallet (Electrum for example) that you can use to monitor your Bitcoin Core balance.
+
+:warning: this means that your Bitcoin Core wallet cannot send funds on its on (since it cannot access private keys to sign transactions).
```
typo: on its on -> on its own

sstone replied:
fixed in https://github.com/ACINQ/eclair/pull/2613/commits/c00ae33f62d53d4d3710f7dc05a273993eb47657

remyers commented on docs/BitcoinCoreKeys.md:
```diff
@@ -0,0 +1,119 @@
+You can also use `eclair-cli getmasterxpub` to get a BIP32 extended public key that you can import into any compatible Bitcoin wallet
+to create a watch-only wallet (Electrum for example) that you can use to monitor your Bitcoin Core balance.
+
+:warning: this means that your Bitcoin Core wallet cannot send funds on its on (since it cannot access private keys to sign transactions).
```
nit: (a hardware wallet -> (on a hardware wallet

remyers commented on docs/BitcoinCoreKeys.md:
```diff
@@ -0,0 +1,119 @@
+
+### Copy the bitcoin core wallet and `eclair-signer.conf`
+
+Copy your wallet file to a new Bitcoin Core node, and load it with `bitcoin-cli loadwallet`. Bitcoin Core make need to scan the blockchain which could take some time.
```
typo: Bitcoin Core make need to scan -> Bitcoin Core may need to scan

sstone replied:
fixed in https://github.com/ACINQ/eclair/pull/2613/commits/c00ae33f62d53d4d3710f7dc05a273993eb47657

remyers commented on docs/Configure.md:
```diff
@@ -4,7 +4,7 @@
 * [Configuration file](#configuration-file)
   * [Changing the data directory](#changing-the-data-directory)
-  * [Splitting the configuration](#splitting-the-configuration)
+  * [Splitting the configuration](#splitting-the-configuration)\
```
is that \ at the end a typo?

sstone replied:
fixed in https://github.com/ACINQ/eclair/pull/2613/commits/c00ae33f62d53d4d3710f7dc05a273993eb47657

sstone commented on eclair-core/src/test/scala/fr/acinq/eclair/integration/IntegrationSpec.scala:
```diff
@@ -146,8 +147,12 @@ abstract class IntegrationSpec extends TestKitBaseClass with BitcoindService wit
     implicit val system: ActorSystem = ActorSystem(s"system-$name", config)
-    val setup = new Setup(datadir, pluginParams = Seq.empty)
+
+    val setup = new Setup(datadir, pluginParams = Seq.empty, seeds_opt = Some(Setup.Seeds(randomBytes32(), randomBytes32())))
```
That parameter can be removed from `Setup()`: it was used for testing but now tests will write a signer configuration file in eclair's data directory when needed and `Setup` will just pick it up

sstone commented on docs/BitcoinCoreKeys.md:
```diff
@@ -0,0 +1,119 @@
+:warning: to backup the private keys of this wallet you must either backup your mnemonic code and passphrase, or backup the `eclair-signer.conf` file in your eclair
+directory (default is `~/.eclair`) along with your channels and node seed files.
+
+:warning: You can also initialise a backup onchain wallet with the same mnemonic code and passphrase (a hardware wallet for example), but be warned that using them may interfere with your node's operations (for example you may end up
```
fixed in https://github.com/ACINQ/eclair/pull/2613/commits/c00ae33f62d53d4d3710f7dc05a273993eb47657

sstone commented on docs/BitcoinCoreKeys.md:
````diff
@@ -0,0 +1,119 @@
+$ bitcoin-cli -named createwallet wallet_name=eclair disable_private_keys=true blank=true descriptors=true load_on_startup=true
+```
+
+4) Import public descriptors generated by Eclair
````
fixed in https://github.com/ACINQ/eclair/pull/2613/commits/c00ae33f62d53d4d3710f7dc05a273993eb47657

remyers reviewed (changes requested):
I ran through the steps in `BitcoinCoreKeys.md` and fixed some issues I found. You can cherry pick them from [482ca5a](https://github.com/remyers/eclair/commit/482ca5a8237722174540f2cee9bf7a84197bc8c4).

Most are just nits/typos, but I also changed:
```sh
$ eclair-cli getdescriptors | jq --raw-output -c | xargs -0 bitcoin-cli -rpcwallet=eclair importdescriptors
```
to two lines:
```sh
$ eclair-cli getdescriptors --account=0 | jq --raw-output -c > descriptors.json
$ cat descriptors.json | xargs -0 bitcoin-cli -rpcwallet=eclair importdescriptors
```
I think this is more how it will be used because Eclair will always have already created/imported the descriptors if `getdescriptors` works. But you might want to migrate to a new instance of bitcoind that can spend time syncing while you continue to use an old bitcoind without blocking.

sstone commented:
> I ran through the steps in `BitcoinCoreKeys.md` and fixed some issues I found. You can cherry pick them from [482ca5a](https://github.com/remyers/eclair/commit/482ca5a8237722174540f2cee9bf7a84197bc8c4).
> 
> Most are just nits/typos, but I also changed:
> 
> ```shell
> $ eclair-cli getdescriptors | jq --raw-output -c | xargs -0 bitcoin-cli -rpcwallet=eclair importdescriptors
> ```
> 
> to two lines:
> 
> ```shell
> $ eclair-cli getdescriptors --account=0 | jq --raw-output -c > descriptors.json
> $ cat descriptors.json | xargs -0 bitcoin-cli -rpcwallet=eclair importdescriptors
> ```
> 
> I think this is more how it will be used because Eclair will always have already created/imported the descriptors if `getdescriptors` works. But you might want to migrate to a new instance of bitcoind that can spend time syncing while you continue to use an old bitcoind without blocking.

I merged your changes and made the `account` parameter optional, with a default value of 0 which is what most users will want to use.

t-bast commented on eclair-core/src/main/scala/fr/acinq/eclair/blockchain/bitcoind/rpc/BitcoinCoreClient.scala:
```diff
@@ -42,13 +46,47 @@ import scala.util.{Failure, Success, Try}
+    onchainKeyManager_opt match {
+      case None => Future.successful(true) // no eclair-backed wallet is configured
+      case Some(onchainKeyManager) if !rpcClient.wallet.contains(onchainKeyManager.wallet) => Future.successful(true) // configured wallet has a different name
+      case Some(onchainKeyManager) if !onchainKeyManager.getDescriptors(0).descriptors.forall(desc => TimestampSecond(desc.timestamp) >= TimestampSecond.now() - 2.hours) =>
```
Why didn't you expose `timestamp` directly on the `OnChainKeyManager`? We shouldn't need the descriptors to just check the timestamp (and it's weird to check all of them when we know the timestamp will be the same)?

t-bast commented on eclair-core/src/main/scala/fr/acinq/eclair/blockchain/bitcoind/rpc/BitcoinCoreClient.scala:
```diff
@@ -42,13 +46,47 @@ import scala.util.{Failure, Success, Try}
+    rpcClient.invoke("importdescriptors", descriptors).collect {
+      case JArray(results) => results.forall(item => {
+        val JBool(success) = item \ "success"
+        val JArray(_) = item \ "warnings"
```
Unused, remove?

sstone replied:
done in https://github.com/ACINQ/eclair/pull/2613/commits/2fd550937780f4efa93b660bda94343b26d68b6f

t-bast commented on eclair-core/src/main/scala/fr/acinq/eclair/blockchain/bitcoind/rpc/BitcoinCoreClient.scala:
```diff
@@ -42,13 +46,47 @@ import scala.util.{Failure, Success, Try}
+      case Some(onchainKeyManager) =>
+        logger.info(s"Creating a new on-chain eclair-backed wallet in bitcoind: ${onchainKeyManager.wallet}")
+        for {
+          _ <- rpcClient.invoke("createwallet", onchainKeyManager.wallet, true, true, "", false, true, true)
```
Can you add the argument names?

sstone replied:
done in https://github.com/ACINQ/eclair/pull/2613/commits/c5bb39f4043b8174ba3f372ae3ce30665523147d

t-bast commented on eclair-core/src/main/scala/fr/acinq/eclair/blockchain/bitcoind/rpc/BitcoinCoreClient.scala:
```diff
@@ -42,13 +46,47 @@ import scala.util.{Failure, Success, Try}
  * The Bitcoin Core client provides some high-level utility methods to interact with Bitcoin Core.
+ *
+ * @param rpcClient             bitcoin core JSON rpc client
+ * @param onchainKeyManager_opt optional onchain key manager. If provided and its wallet name matcher our rpcClient wallet's name, it will be used to sign transactions (it is assumed that bitcoin
```
nit:

```suggestion
 * @param onchainKeyManager_opt optional onchain key manager. If provided and its wallet name matches our rpcClient wallet's name, it will be used to sign transactions (it is assumed that bitcoin
```

sstone replied:
done in https://github.com/ACINQ/eclair/pull/2613/commits/c5bb39f4043b8174ba3f372ae3ce30665523147d

t-bast reviewed (commented):
A first pass to warm up with the simple parts of the PR: documentation and creation of the eclair-backed wallet.

t-bast commented on eclair-core/src/main/scala/fr/acinq/eclair/channel/publish/ReplaceableTxFunder.scala:
```diff
@@ -312,13 +315,13 @@ private class ReplaceableTxFunder(nodeParams: NodeParams,
   }
 
-  private def sign(fundedTx: ReplaceableTxWithWitnessData, txFeerate: FeeratePerKw, amountIn: Satoshi): Behavior[Command] = {
+  private def sign(fundedTx: ReplaceableTxWithWitnessData, txFeerate: FeeratePerKw, amountIn: Satoshi, packageWeight: Int): Behavior[Command] = {
```
I don't think you need this `packageWeight` argument, and even if you did this wouldn't work because the weight you're passing in doesn't account for the witnesses. I believe you can simply use the weight of the signed tx once you have it, which reduces the number of changes to this file.

sstone replied:
We need to re-compute the weight that was used to fund the transaction (which may include the weight of the tx it wants to bump, using the weight of the signed tx is not enough) but it can be recomputed locally instead of being passed around, see https://github.com/ACINQ/eclair/pull/2613/commits/f9fa8a7da11605335a0610c0dd61b3dcfca36fa5

t-bast commented on eclair-core/src/main/scala/fr/acinq/eclair/channel/publish/ReplaceableTxPrePublisher.scala:
```diff
@@ -64,22 +64,33 @@ object ReplaceableTxPrePublisher {
     def updateTx(tx: Transaction): ReplaceableTxWithWitnessData
   }
-  /** Replaceable transaction for which we may need to add wallet inputs. */
+  /** Replaceable transaction for which we may need to add wallet inputs and outputs. */
```
All the changes made to this file seem unnecessary: for all of these transactions, we have exactly one input and one output that don't belong to the wallet (and must use the same index for HTLC txs because our peer used `SIGHASH_SINGLE | SIGASH_ANYONECANPAY`). We shouldn't need to explicitly track anything, we can do this on-the-fly when evaluating the final transaction.

sstone replied:
Explicit tracking of our wallet inputs and outputs is a small change that makes reasoning about the code simpler (since the PSBT signing code needs to verify such inputs and outputs) and has very little impact on the rest f the code base, I' d rather keep it unless you really think that it should be removed (I agree that this "eclair manages bitcoin keys" mode will not be used very often and we should minimize its impact on our code base, maybe I've missed something...).

t-bast replied:
> Explicit tracking of our wallet inputs and outputs is a small change that makes reasoning about the code simpler

I disagree...I think it's unnecessary complexity: it adds quite a lot of code that isn't easy at all to reason about (since those values are set and read across different components that make regular changes to the structure of the transactions and need to make sure they also correctly update those new fields). The various transaction publishers are already quite complex (mostly because we need to work around some bitcoind limitations, which I hope we can simplify in the future) and are extremely sensitive, so I'd rather minimize changes and optimize for simplicity. Always using the first position in the transaction for our non-wallet input and output is the simplest thing I can think of, don't you agree?

sstone replied:
Changed in https://github.com/ACINQ/eclair/pull/2613/commits/aecae9aded205cb0bbacca99a3a23b520821fb4e

t-bast commented on eclair-core/src/main/scala/fr/acinq/eclair/channel/fund/InteractiveTxBuilder.scala:
```diff
@@ -766,16 +766,30 @@ private class InteractiveTxBuilder(replyTo: ActorRef[InteractiveTxBuilder.Respon
-        case SignTransactionResponse(signedTx, _) =>
+      // we only sign our wallet inputs, and check that we can spend our wallet outputs
+      val ourWalletInputs = unsignedTx.localInputs.map(i => tx.txIn.indexWhere(_.outPoint == OutPoint(i.previousTx, i.previousTxOutput.toInt)))
+      val ourWalletOutputs = unsignedTx.localOutputs.map(o => tx.txOut.indexWhere(output => output.amount == o.amount && output.publicKeyScript == o.pubkeyScript))
```
The `localOutputs` can contain either `Local.Change` or `Local.NonChange` (splice-out). If you use all of them here, that means we restrict splice-out to only send to our own wallet. Is that what we want? I believe users will want to be able to splice-out to external addresses.

We definitely want our node to have this restriction, we should make this configurable? We should add to `eclair.conf` a boolean `allow-splice-out-external-addresses` and tweak our behavior here based on this flag.

sstone replied:
I think that skipping verification of `Local.NonChange` should work if we assume that such outputs have been set by trusted sources (like a secure API call), see https://github.com/ACINQ/eclair/pull/2613/commits/5b4572e8a04de991fe090dc3ceadd813deb0c69a, wdyt ?

t-bast replied:
That sounds reasonable indeed and should work for our current use-cases :+1:

t-bast commented on eclair-core/src/main/scala/fr/acinq/eclair/crypto/keymanager/LocalOnchainKeyManager.scala:
```diff
@@ -0,0 +1,165 @@
+    ))
+  }
+
+  override def signPsbt(psbt: Psbt, ourInputs: Seq[Int], ourOutputs: Seq[Int]): Psbt = {
```
This should be explicit and return a `Try[Psbt]`: there are a lot of ways this can fail, so the caller really needs to know about it.

I understand why you chose to rely on exceptions (because this is only called inside a future that we want to fail if `signPsbt` didn't succeed), but it makes the code of some of the helper functions below hard to reason about, some of them can be simplified by refactoring to avoid exceptions (I'll give this a try in my own branch).

sstone replied:
done in https://github.com/ACINQ/eclair/pull/2613/commits/a53c5ec8098e9f80da83ba72278fb2d67783b148

t-bast commented on eclair-core/src/main/scala/fr/acinq/eclair/channel/publish/ReplaceableTxFunder.scala:
```diff
@@ -341,31 +344,52 @@ private class ReplaceableTxFunder(nodeParams: NodeParams,
+    // we finalize (sign) the input that we control, and will then ask our bitcoin client to sign wallet inputs
+    val psbt = new Psbt(locallySignedTx.txInfo.tx)
+    val updated = psbt.updateWitnessInput(locallySignedTx.txInfo.input.outPoint, locallySignedTx.txInfo.input.txOut, null, fr.acinq.bitcoin.Script.parse(locallySignedTx.txInfo.input.redeemScript), null, java.util.Map.of())
+    val finalized = EitherKt.flatMap(updated, (psbt: Psbt) => psbt.finalizeWitnessInput(0, locallySignedTx.txInfo.tx.txIn.head.witness))
```
Dealing with a mix of kotlin `Either` and scala `Either` is anoying, can't we have an automatic conversion in our `KotlinUtils`? It would be easier to read if we had more idiomatic scala here.

sstone replied:
done in https://github.com/ACINQ/eclair/pull/2613/commits/05650d44272d7ed480c946274ad1bb5be2fe7a90

t-bast commented on eclair-core/src/main/scala/fr/acinq/eclair/channel/publish/ReplaceableTxFunder.scala:
```diff
@@ -341,31 +344,52 @@ private class ReplaceableTxFunder(nodeParams: NodeParams,
+        case Success(processPsbtResponse) =>
+          val signedTx = processPsbtResponse.finalTx
+          val actualFees = kmp2scala(processPsbtResponse.psbt.computeFees())
+          val actualFeerate = FeeratePerKw((actualFees * 1000) / packageWeight)
```
Why not just `Transactions.fee2rate(actualFees, signedTx.weight())`?

sstone replied:
done in https://github.com/ACINQ/eclair/pull/2613/commits/2fd550937780f4efa93b660bda94343b26d68b6f

t-bast reviewed (commented):
Concept ACK, the scope of this PR looks good, and you've been able to reduce it to a somewhat small change (mostly just a move to using the PSBT RPCs and verifying the results bitcoind sends back).

I think it can be simplified further (see comments below), but otherwise this looks good. I think this will deserve some clean-up to remove some kotlin<->scala awkwardness and make some functions more readable, but this is a good start!

sstone commented on docs/BitcoinCoreKeys.md:
```diff
@@ -0,0 +1,120 @@
+
+You can use any BIP39-compatible tool, including most hardware wallets.
+
+2) Create an `eclair-signer.conf` configuration file add it to eclair's data directory
```
Using a new `signer` section is `eclair.conf` is simpler but "leaks" because it becomes available everywhere in the code through the actor system's settings instead of being contained to where it is actually needed and could potentially be exposed through a bug that "exports" the configuration (through logs, ....) through this is highly unlikely (especially if eclair runs in a environment where we explicitly control all I/O).

I don't feel very strongly about this and using a new section in `eclair.conf` would be a very simple change though.

t-bast replied:
Those are good reasons to keep it separate :+1: , it's maybe worth a comment in the code to explain the rationale

sstone commented on docs/BitcoinCoreKeys.md:
```diff
@@ -0,0 +1,120 @@
+# Using Eclair to manage your Bitcoin Core wallet's private keys
+
+You can configure Eclair to control (and never expose) the private keys of your Bitcoin Core wallet. This feature was designed to take advantage of deployements where your Eclair node runs in a
```
done in https://github.com/ACINQ/eclair/pull/2613/commits/c5bb39f4043b8174ba3f372ae3ce30665523147d

sstone commented on docs/BitcoinCoreKeys.md:
```diff
@@ -0,0 +1,120 @@
+:warning: to backup the private keys of this wallet you must either backup your mnemonic code and passphrase, or backup the `eclair-signer.conf` file in your eclair
+directory (default is `~/.eclair`) along with your channels and node seed files.
+
+:warning: You can also initialise a backup onchain wallet with the same mnemonic code and passphrase (on a hardware wallet for example), but be warned that using them may interfere with your node's operations (for example you may end up
```
done in https://github.com/ACINQ/eclair/pull/2613/commits/c5bb39f4043b8174ba3f372ae3ce30665523147d

t-bast replied:
This doesn't seem to be fixed in the latest diff?

sstone replied:
Ah sorry I would argue that `initialise` is the proper spelling but I changed it in https://github.com/ACINQ/eclair/pull/2613/commits/6d946a85b072400d07f2c46e20dcde700da37e47

t-bast replied:
Oh I didn't know that it could also be used (I guess UK vs US?), my auto-correct highlighted it, that's why I reacted.

sstone commented on eclair-core/src/test/scala/fr/acinq/eclair/router/AnnouncementsBatchValidationSpec.scala:
```diff
@@ -85,7 +86,7 @@ object AnnouncementsBatchValidationSpec {
   }
 
-  def simulateChannel(bitcoinClient: BitcoinCoreClient)(implicit ec: ExecutionContext): SimulatedChannel = {
+  def simulateChannel(bitcoinClient: BitcoinCoreClient, onchainKeyManager: OnchainKeyManager)(implicit ec: ExecutionContext): SimulatedChannel = {
```
done in https://github.com/ACINQ/eclair/pull/2613/commits/c5bb39f4043b8174ba3f372ae3ce30665523147d

sstone commented on eclair-core/src/main/scala/fr/acinq/eclair/Setup.scala:
```diff
@@ -139,13 +141,27 @@ class Setup(val datadir: File,
       port = config.getInt("bitcoind.rpcport"),
       wallet = wallet)
+
+    def createDescriptorWallet(wallets: List[String]): Future[Boolean] = {
```
done in https://github.com/ACINQ/eclair/pull/2613/commits/c5bb39f4043b8174ba3f372ae3ce30665523147d

sstone commented on eclair-core/src/test/scala/fr/acinq/eclair/integration/IntegrationSpec.scala:
```diff
@@ -146,8 +147,12 @@ abstract class IntegrationSpec extends TestKitBaseClass with BitcoindService wit
     implicit val system: ActorSystem = ActorSystem(s"system-$name", config)
-    val setup = new Setup(datadir, pluginParams = Seq.empty)
+
+    val setup = new Setup(datadir, pluginParams = Seq.empty, seeds_opt = Some(Setup.Seeds(randomBytes32(), randomBytes32())))
```
done in https://github.com/ACINQ/eclair/pull/2613/commits/c5bb39f4043b8174ba3f372ae3ce30665523147d

sstone commented on eclair-core/src/test/scala/fr/acinq/eclair/integration/ChannelIntegrationSpec.scala:
```diff
@@ -455,9 +455,12 @@ abstract class ChannelIntegrationSpec extends IntegrationSpec {
-    instantiateEclairNode("A", ConfigFactory.parseMap(Map("eclair.node-alias" -> "A", "eclair.channel.expiry-delta-blocks" -> 40, "eclair.channel.fulfill-safety-before-timeout-blocks" -> 12, "eclair.server.port" -> 29740, "eclair.api.port" -> 28090).asJava).withFallback(withDefaultCommitment).withFallback(commonConfig))
-    instantiateEclairNode("C", ConfigFactory.parseMap(Map("eclair.node-alias" -> "C", "eclair.channel.expiry-delta-blocks" -> 40, "eclair.channel.fulfill-safety-before-timeout-blocks" -> 12, "eclair.server.port" -> 29741, "eclair.api.port" -> 28091).asJava).withFallback(withAnchorOutputs).withFallback(commonConfig))
-    instantiateEclairNode("F", ConfigFactory.parseMap(Map("eclair.node-alias" -> "F", "eclair.channel.expiry-delta-blocks" -> 40, "eclair.channel.fulfill-safety-before-timeout-blocks" -> 12, "eclair.server.port" -> 29742, "eclair.api.port" -> 28092).asJava).withFallback(withDefaultCommitment).withFallback(commonConfig))
+    val mapA = Map("eclair.node-alias" -> "A", "eclair.channel.expiry-delta-blocks" -> 40, "eclair.channel.fulfill-safety-before-timeout-blocks" -> 12, "eclair.server.port" -> TestUtils.availablePort, "eclair.api.port" -> TestUtils.availablePort)
```
done in https://github.com/ACINQ/eclair/pull/2613/commits/c5bb39f4043b8174ba3f372ae3ce30665523147d

sstone commented on eclair-core/src/test/scala/fr/acinq/eclair/integration/ChannelIntegrationSpec.scala:
```diff
@@ -455,9 +455,12 @@ abstract class ChannelIntegrationSpec extends IntegrationSpec {
-    instantiateEclairNode("C", ConfigFactory.parseMap(Map("eclair.node-alias" -> "C", "eclair.channel.expiry-delta-blocks" -> 40, "eclair.channel.fulfill-safety-before-timeout-blocks" -> 12, "eclair.server.port" -> 29741, "eclair.api.port" -> 28091).asJava).withFallback(withAnchorOutputs).withFallback(commonConfig))
-    instantiateEclairNode("F", ConfigFactory.parseMap(Map("eclair.node-alias" -> "F", "eclair.channel.expiry-delta-blocks" -> 40, "eclair.channel.fulfill-safety-before-timeout-blocks" -> 12, "eclair.server.port" -> 29742, "eclair.api.port" -> 28092).asJava).withFallback(withDefaultCommitment).withFallback(commonConfig))
+    val mapA = Map("eclair.node-alias" -> "A", "eclair.channel.expiry-delta-blocks" -> 40, "eclair.channel.fulfill-safety-before-timeout-blocks" -> 12, "eclair.server.port" -> TestUtils.availablePort, "eclair.api.port" -> TestUtils.availablePort)
+    val mapB = Map("eclair.node-alias" -> "C", "eclair.channel.expiry-delta-blocks" -> 40, "eclair.channel.fulfill-safety-before-timeout-blocks" -> 12, "eclair.server.port" -> TestUtils.availablePort, "eclair.api.port" -> TestUtils.availablePort)
```
done in https://github.com/ACINQ/eclair/pull/2613/commits/c5bb39f4043b8174ba3f372ae3ce30665523147d

t-bast reviewed (commented):
I spent a lot of time staring at the code, playing with it and tweaking it, it looks mostly good to me!
All my comments are addressed by a follow-up PR at #2726, once this is integrated this feel ready :+1:

t-bast reviewed (approved):
LGTM :rocket:

t-bast commented:
LGTM, I'll start working on a follow-up PR to simplify the `ReplaceableTxFunder`

pm47 commented on eclair-core/src/main/scala/fr/acinq/eclair/channel/fund/InteractiveTxBuilder.scala:
```diff
@@ -766,16 +766,39 @@ private class InteractiveTxBuilder(replyTo: ActorRef[InteractiveTxBuilder.Respon
     } else {
-      context.pipeToSelf(wallet.signTransaction(tx, allowIncomplete = true).map {
-        case SignTransactionResponse(signedTx, _) =>
+      // We only sign our wallet inputs, and check that we can spend our wallet outputs.
```
AFAICT we don't "check that we can spend our wallet outputs.", we rather verify that the outputs that we have specified have been left untouched? In that case, why does it matter that we control them or not (change vs splice output)?

pm47 replied:
Let me try to rephrase. I think we should:
- verify that all provided inputs and outputs have been preserved (entirely: amounts, scripts, etc.)
- verify that we control all new output scripts

pm47 reviewed (commented):
This looks mostly good, and pretty well contained given the significance of the change. I have only minor comments.

My main remark would be on the structure of the PR: the PSBT change should have been made an independent commit at the beginning. It would have made it much easier to assess the impact of the delegated signature, which is the core of the PR.

I'm not familiar with PSBT and haven't reviewed the related code.

sstone commented on eclair-core/src/main/scala/fr/acinq/eclair/crypto/keymanager/OnChainKeyManager.scala:
```diff
@@ -0,0 +1,70 @@
+import scala.util.Try
+
+trait OnChainKeyManager {
+  def wallet: String
```
done in https://github.com/ACINQ/eclair/pull/2613/commits/0f47b52ebcfdb1b8e232d203783e3455ac90d82a

sstone commented on eclair-core/src/main/scala/fr/acinq/eclair/Setup.scala:
```diff
@@ -305,7 +314,7 @@ class Setup(val datadir: File,
       _ <- Future.firstCompletedOf(routerInitialized.future :: routerTimeout :: Nil)
 
-      _ = bitcoinClient.getReceiveAddress().map(address => logger.info(s"initial wallet address=$address"))
+      _ = bitcoinClient.getReceiveAddress().map(address => logger.info(s"initial address=$address for bitcoin wallet=${bitcoinClient.rpcClient.wallet.getOrElse("")}"))
```
done in https://github.com/ACINQ/eclair/pull/2613/commits/0f47b52ebcfdb1b8e232d203783e3455ac90d82a

sstone commented on eclair-core/src/main/scala/fr/acinq/eclair/Setup.scala:
```diff
@@ -244,7 +252,7 @@ class Setup(val datadir: File,
       finalPubkey = new AtomicReference[PublicKey](null)
       pubkeyRefreshDelay = FiniteDuration(config.getDuration("bitcoind.final-pubkey-refresh-delay").getSeconds, TimeUnit.SECONDS)
-      bitcoinClient = new BitcoinCoreClient(bitcoin) with OnchainPubkeyCache {
+      bitcoinClient = new BitcoinCoreClient(bitcoin, if (bitcoin.wallet == onChainKeyManager_opt.map(_.wallet)) onChainKeyManager_opt else None) with OnchainPubkeyCache {
```
We need this to load an onchain key manager that cab be used to generate descriptors through the API but is not active because wallet names in `eclair.conf` and `eclair-signer.conf` are different.

pm47 replied:
This is not at all obvious and deserves a comment.

I think this may also clash with #2737 but we can figure this out when we merge that PR.

sstone commented on eclair-core/src/main/scala/fr/acinq/eclair/crypto/keymanager/LocalOnChainKeyManager.scala:
```diff
@@ -0,0 +1,239 @@
+import scala.util.{Failure, Success, Try}
+
+object LocalOnChainKeyManager extends Logging {
+  def descriptorChecksum(span: String): String = fr.acinq.bitcoin.Descriptor.checksum(span)
```
I've removed it: it is provided by `bitcoin-kmp` now and internal to how we generate descriptors

sstone commented on docs/BitcoinCoreKeys.md:
```diff
@@ -0,0 +1,122 @@
+### 3. Configure Eclair to handle private keys for this wallet
+
+Set `eclair.bitcoind.wallet` to the name of the wallet in your `eclair-signer.conf` file (`eclair` in the example above) and restart Eclair.
+Eclair will automatically create a new, empty, descriptor-enabled, watch-only wallet in Bitcoin Core and import its descriptors.
```
`blank=true` means that the wallet will initially be empty, but it will will possible to import HD seeds or descriptors later to create private keys.
`disable_private_keys=true` means that it will not be possible to generate private keys, and this is what makes the wallet watch-only.

t-bast reviewed (dismissed):
LGTM, great job I like how small the change ends up being compared to what it accomplishes. The test that randomly fails is unrelated to this PR, it's on my TODO list (it's related to the migration to bitcoind 24.1).

pm47 reviewed (approved):
LGTM 🔑 

There is some polishing left on the doc, and on namings (there are two many related terms: clair signer ? managed bitcoin keys? onchain key manager?) but that can be done in a follow up PR.
